core/src/test/java/feast/core/service/SpecServiceTest.java

```diff
@@ -367,6 +355,100 @@ public void applyFeatureSetShouldNotCreateFeatureSetIfFieldsUnordered()
     assertThat(applyFeatureSetResponse.getFeatureSet().getName(), equalTo(expected.getName()));
   }
 
+  @Test
+  public void applyFeatureSetShouldAcceptFeatureLabels() throws InvalidProtocolBufferException {
+    List<FeatureSetProto.EntitySpec> entitySpecs = new ArrayList<>();
+    entitySpecs.add(
+        FeatureSetProto.EntitySpec.newBuilder()
+            .setName("entity1")
+            .setValueType(Enum.INT64)
+            .build());
+
+    Map<String, String> featureLabels0 =
+        new HashMap<String, String>() {
+          {
+            put("label1", "feast1");
+          }
+        };
+
+    Map<String, String> featureLabels1 =
+        new HashMap<String, String>() {
+          {
+            put("label1", "feast1");
+            put("label2", "feast2");
+          }
+        };
+
+    List<Map<String, String>> featureLabels = new ArrayList<>();
+    featureLabels.add(featureLabels0);
+    featureLabels.add(featureLabels1);
+
+    List<FeatureSpec> featureSpecs = new ArrayList<>();
+    featureSpecs.add(
+        FeatureSpec.newBuilder()
+            .setName("feature1")
+            .setValueType(Enum.INT64)
+            .putAllLabels(featureLabels.get(0))
+            .build());
+    featureSpecs.add(
+        FeatureSpec.newBuilder()
+            .setName("feature2")
+            .setValueType(Enum.INT64)
+            .putAllLabels(featureLabels.get(1))
+            .build());
+
+    FeatureSetSpec featureSetSpec =
+        FeatureSetSpec.newBuilder()
+            .setName("featureSetWithConstraints")
+            .addAllEntities(entitySpecs)
+            .addAllFeatures(featureSpecs)
+            .build();
+
+    ApplyFeatureSetResponse applyFeatureSetResponse = specService.applyFeatureSet(featureSetSpec);
+    FeatureSetSpec appliedFeatureSetSpec = applyFeatureSetResponse.getFeatureSet();
+
+    // appliedEntitySpecs needs to be sorted because the list returned by specService may not
+    // follow the order in the request
+    List<FeatureSetProto.EntitySpec> appliedEntitySpecs =
+        new ArrayList<>(appliedFeatureSetSpec.getEntitiesList());
+    appliedEntitySpecs.sort(Comparator.comparing(EntitySpec::getName));
+
+    // appliedFeatureSpecs needs to be sorted because the list returned by specService may not
+    // follow the order in the request
+    List<FeatureSpec> appliedFeatureSpecs =
+        new ArrayList<>(appliedFeatureSetSpec.getFeaturesList());
+    appliedFeatureSpecs.sort(Comparator.comparing(FeatureSpec::getName));
+
+    List<Map<String, String>> featureSpecsLabels =
+        featureSpecs.stream().map(e -> e.getLabelsMap()).collect(Collectors.toList());
+    assertThat(appliedEntitySpecs, equalTo(entitySpecs));
+    assertThat(appliedFeatureSpecs, equalTo(featureSpecs));
+    assertThat(featureSpecsLabels, equalTo(featureLabels));
```

Comment:
I just noticed this when raising this PR: I believe that the `featureSpecsLabels` here should be assigned from `appliedFeatureSpecs.stream()` instead of `featureSpecs.stream()`, otherwise we are asserting on test setup code instead of covering actual roundtrip through the feature set being applied. I've asked @suwik to double-check my reasoning.

I've made that change locally and the test passes, so I believe it proves there is no functional issue with feature-level labels, we just aren't covering what we want to with the test. I'll include the update in this PR if @suwik corroborates, but the same change should be forward-ported to master, so I want to track that.